core/src/main/java/com/linecorp/armeria/server/streaming/ServerSentEvents.java

```diff
@@ -0,0 +1,341 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server.streaming;
+
+import static com.linecorp.armeria.internal.ResponseConversionUtil.streamingFrom;
+import static com.linecorp.armeria.server.streaming.SanitizationUtil.ensureContentType;
+import static com.linecorp.armeria.server.streaming.SanitizationUtil.ensureHttpStatus;
+import static java.util.Objects.requireNonNull;
+
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+import org.reactivestreams.Publisher;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.common.sse.ServerSentEvent;
+
+/**
+ * A utility class which helps to create a <a href="https://www.w3.org/TR/eventsource/">Server-sent Events</a>
+ * stream from a content {@link Publisher} or {@link Stream}.
+ *
+ * <p>A user simply creates a streaming {@link HttpResponse} which emits Server-sent Events, e.g.
+ * <pre>{@code
+ * Server server = new ServerBuilder()
+ *         // Emit Server-sent Events with the SeverSentEvent instances published by a publisher.
+ *         .service("/sse1",
+ *                  (ctx, req) -> ServerSentEvents.fromPublisher(
+ *                          Flux.just(ServerSentEvent.ofData("foo"), ServerSentEvent.ofData("bar"))))
+ *         // Emit Server-sent Events with converting instances published by a publisher into
+ *         // ServerSentEvent instances.
+ *         .service("/sse2",
+ *                  (ctx, req) -> ServerSentEvents.fromPublisher(
+ *                          Flux.just("foo", "bar"), ServerSentEvent::ofData))
+ *         .build();
+ * }</pre>
+ */
+public final class ServerSentEvents {
+
+    /**
+     * A line feed character which marks the end of a field in Server-sent Events.
+     */
+    private static final char LINE_FEED = '\n';
+
+    /**
+     * A default {@link HttpHeaders} of Server-sent Events.
+     */
+    private static final HttpHeaders defaultHttpHeaders =
+            HttpHeaders.of(HttpStatus.OK).contentType(MediaType.EVENT_STREAM).asImmutable();
+
+    /**
+     * Creates a new Server-sent Events stream from the specified {@link Publisher}.
+     *
+     * @param contentPublisher the {@link Publisher} which publishes the objects supposed to send as contents
+     */
+    public static <T extends ServerSentEvent<?>> HttpResponse fromPublisher(Publisher<T> contentPublisher) {
```

Comment:
Global comment: In this case, you can just use `Publisher<? extends ServerSentEvent>`.

Reply:
Fixed. Thanks!